src/frontend/src/optimizer/rule/top_n_on_index_rule.rs

```diff
@@ -0,0 +1,133 @@
+use crate::optimizer::property::{Direction, FieldOrder, Order};
+use crate::optimizer::PlanRef;
+
+pub struct TopNOnIndexRule {}
+
+impl Rule for TopNOnIndexRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let logical_top_n: &LogicalTopN = plan.as_logical_top_n()?;
+        let logical_scan: LogicalScan = logical_top_n.input().as_logical_scan()?.to_owned();
+        let order = logical_top_n.topn_order();
+        if order.field_order.is_empty() {
+            return None;
+        }
+        if let Some(p) = self.try_on_pk(logical_top_n, logical_scan.clone(), order) {
+            Some(p)
+        } else {
+            self.try_on_index(logical_top_n, logical_scan, order)
+        }
+    }
+}
+
+impl TopNOnIndexRule {
+    pub fn create() -> BoxedRule {
+        Box::new(TopNOnIndexRule {})
+    }
+
+    fn try_on_index(
+        &self,
+        logical_top_n: &LogicalTopN,
+        logical_scan: LogicalScan,
+        order: &Order,
+    ) -> Option<PlanRef> {
+        let index = logical_scan.indexes().iter().find(|idx| {
+            Order {
+                field_order: idx
+                    .index_table
+                    .pk()
+                    .iter()
+                    .map(|idx_item| FieldOrder {
+                        index: idx_item.index,
+                        direct: idx_item.direct,
+                    })
+                    .collect(),
+            }
+            .satisfies(order)
+        })?;
+
+        let p2s_mapping = index.primary_to_secondary_mapping();
+
+        let mut index_scan = if logical_scan
+            .required_col_idx()
+            .iter()
+            .all(|x| p2s_mapping.contains_key(x))
+        {
+            Some(logical_scan.to_index_scan(
+                &index.name,
+                index.index_table.table_desc().into(),
+                p2s_mapping,
+            ))
+        } else {
+            None
+        }?;
+
+        index_scan.set_chunk_size(
+            ((u32::MAX as u64).min(logical_top_n.limit() + logical_top_n.offset())) as usize,
+        );
+
+        let logical_limit = LogicalLimit::create(
+            index_scan.into(),
+            logical_top_n.limit(),
+            logical_top_n.offset(),
+        );
+        Some(logical_limit)
+    }
+
+    fn try_on_pk(
+        &self,
+        logical_top_n: &LogicalTopN,
+        mut logical_scan: LogicalScan,
+        order: &Order,
+    ) -> Option<PlanRef> {
+        let primary_key = logical_scan.get_primary_key();
+        let primary_key_order = Order {
+            field_order: primary_key
+                .into_iter()
+                .map(|op| FieldOrder {
+                    index: op.column_idx,
+                    direct: if op.order_type == OrderType::Ascending {
+                        Direction::Asc
+                    } else {
+                        Direction::Desc
+                    },
```

Comment:
😄 I guess @richardchien is working on unifying these types.

Reply:
:doge:

BTW there's no conflict for now. I'll unify these types after finishing distinct aggregator, which may need some days.